managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -558,76 +591,193 @@ protected void recoverFromLedger(final ManagedCursorInfo info, final VoidCallbac
+                        recoverFromLedgerByEntryId(info, callback, lh, entryId - 1);
+                    }
                 }
+            } catch (IOException error) {
+                log.error("Cannot parse footer", error);
+                log.warn("[{}] Error recovering from metadata ledger {} entry {} for cursor {}, cannot parse footer. "
+                                + "Will try recovery from previous entry.",
+                        ledger.getName(), ledgerId, entryId, name, error);
+                recoverFromLedgerByEntryId(info, callback, lh, entryId - 1);
+            }
+        }, null);
+    }
 
-                LedgerEntry entry = seq.nextElement();
-                mbean.addReadCursorLedgerSize(entry.getLength());
-                PositionInfo positionInfo;
-                try {
-                    positionInfo = PositionInfo.parseFrom(entry.getEntry());
-                } catch (InvalidProtocolBufferException e) {
-                    callback.operationFailed(new ManagedLedgerException(e));
-                    return;
-                }
+    private void readChunkSequence(VoidCallback callback, LedgerHandle lh,
+                                   long footerPosition, ChunkSequenceFooter chunkSequenceFooter) {
+        long startPos = footerPosition - chunkSequenceFooter.numParts;
+        long endPos = footerPosition - 1;
+        log.info("readChunkSequence from pos {}, num parts {}, startPos {}, endPos {}",
+                footerPosition, chunkSequenceFooter.numParts, startPos, endPos);
+        lh.asyncReadEntries(startPos, endPos, new AsyncCallback.ReadCallback() {
+            @Override
+            public void readComplete(int rc, LedgerHandle lh, Enumeration<LedgerEntry> entries, Object ctx) {
+                CompositeByteBuf buffer = PulsarByteBufAllocator.DEFAULT.compositeBuffer();
 
-                Map<String, Long> recoveredProperties = Collections.emptyMap();
-                if (positionInfo.getPropertiesCount() > 0) {
-                    // Recover properties map
-                    recoveredProperties = new HashMap<>();
-                    for (int i = 0; i < positionInfo.getPropertiesCount(); i++) {
-                        LongProperty property = positionInfo.getProperties(i);
-                        recoveredProperties.put(property.getName(), property.getValue());
+                AtomicInteger readableBytes = new AtomicInteger(0);
+                entries.asIterator().forEachRemaining(entry -> {
+                    if (log.isInfoEnabled()) {
+                        log.debug("pos {} len {} bytes ", entry.getEntryId(), entry.getLength());
                     }
+                    ByteBuf part = entry.getEntryBuffer();
+                    buffer.addComponent(part);
+                    readableBytes.addAndGet(part.readableBytes());
+                });
+                buffer.writerIndex(readableBytes.get())
+                        .readerIndex(0);
```

Comment:
instead of adjusting the writerIndex manually, `buffer.addComponent(true, part)` could be a way to achieve this.